Connector builder: Port "send stacktrace when error on read" to CDK connector builder module

girarda opened the pull request:
## What
* I forgot to port [this PR](https://github.com/airbytehq/airbyte-platform-internal/pull/5108/) which was in flight when I [ported over the grouping logic](https://github.com/airbytehq/airbyte/pull/23925)
* On uncaught error, stop processing message group and return the error with the stacktrace as an `AirbyteLogMessage`. 
* The connector builder handler will wrap the error log message in a StreamRead so the FE can render it

## How
* Port over the ErrorFormatter and the LogMessage dataclass
* Return errors as LogMessages instead of TracedException
* Only validate pages have a request and a response if the Source encountered no errors

## Recommended reading order
1. `airbyte-cdk/python/connector_builder/utils/error_formatter.py`
2. `airbyte-cdk/python/connector_builder/models.py`
3. `airbyte-cdk/python/connector_builder/connector_builder_handler.py`
4. `airbyte-cdk/python/connector_builder/message_grouper.py`

## Sample output record from a request that did not result in a response
Showing this sample to show the error isn't swallowed by the page validation:
```
{
  "type": "RECORD",
  "record": {
    "stream": "Accounts",
    "data": {
      "logs": [
        {
          "message": "slice:{\"start_time\": \"2022-01-01\", \"end_time\": \"2022-01-01\"}",
          "level": "INFO"
        },
        {
          "message": "HTTPSConnectionPool(host='localhost', port=9999): Max retries exceeded with url: /accounts (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x105ad4850>: Failed to establish a new connection: [Errno 61] Connection refused')) - Traceback (most recent call last):\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-gnews/.venv/lib/python3.9/site-packages/urllib3/connection.py\", line 174, in _new_conn\n    conn = connection.create_connection(\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-gnews/.venv/lib/python3.9/site-packages/urllib3/util/connection.py\", line 95, in create_connection\n    raise err\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-gnews/.venv/lib/python3.9/site-packages/urllib3/util/connection.py\", line 85, in create_connection\n    sock.connect(sa)\nConnectionRefusedError: [Errno 61] Connection refused\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-gnews/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 703, in urlopen\n    httplib_response = self._make_request(\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-gnews/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 386, in _make_request\n    self._validate_conn(conn)\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-gnews/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 1042, in _validate_conn\n    conn.connect()\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-gnews/.venv/lib/python3.9/site-packages/urllib3/connection.py\", line 358, in connect\n    self.sock = conn = self._new_conn()\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-gnews/.venv/lib/python3.9/site-packages/urllib3/connection.py\", line 186, in _new_conn\n    raise NewConnectionError(\nurllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPSConnection object at 0x105ad4850>: Failed to establish a new connection: [Errno 61] Connection refused\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-gnews/.venv/lib/python3.9/site-packages/requests/adapters.py\", line 489, in send\n    resp = conn.urlopen(\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-gnews/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 787, in urlopen\n    retries = retries.increment(\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-gnews/.venv/lib/python3.9/site-packages/urllib3/util/retry.py\", line 592, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='localhost', port=9999): Max retries exceeded with url: /accounts (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x105ad4850>: Failed to establish a new connection: [Errno 61] Connection refused'))\n\nDuring handling of t:
...
```

## Sample output record from a request resulting in a 404

```
{
  "type": "RECORD",
  "record": {
    "stream": "Accounts",
    "data": {
      "logs": [
        {
          "message": "slice:{\"start_time\": \"2022-01-01\", \"end_time\": \"2022-01-01\"}",
          "level": "INFO"
        },
        {
          "message": "Request to https://api.timelyapp.com/1./accounts failed with status code 404 and error message None - Traceback (most recent call last):\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/connector_builder/message_grouper.py\", line 144, in _read_stream\n    yield from source.read(logger=self.logger, config=config, catalog=configured_catalog, state={})\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/declarative/manifest_declarative_source.py\", line 119, in read\n    yield from super().read(logger, config, catalog, state)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py\", line 134, in read\n    raise e\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py\", line 120, in read\n    yield from self._read_stream(\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py\", line 189, in _read_stream\n    for record in record_iterator:\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py\", line 307, in _read_full_refresh\n    for record_data_or_message in record_data_or_messages:\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_stream.py\", line 100, in read_records\n    for record in self.retriever.read_records(sync_mode, cursor_field, stream_slice, stream_state):\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py\", line 375, in read_records\n    for record in records_generator:\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py\", line 440, in _read_pages\n    yield from records_generator_fn(request, response, stream_state, stream_slice)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py\", line 416, in _parse_records_and_emit_request_and_responses\n    yield from self.parse_response(response, stream_slice=stream_slice, stream_state=stream_state)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py\", line 327, in parse_response\n    raise ReadException(error_message)\nairbyte_cdk.sources.declarative.exceptions.ReadException: Request to https://api.timelyapp.com/1./accounts failed with status code 404 and error message None\n",
          "level": "ERROR"
        }
      ],
"slices": [
...
]
...
}
```

maxi297 commented:
I can't give approval right now. I might not understand the full loop but it seems like this change is missing : https://github.com/airbytehq/airbyte-platform-internal/pull/5108/files#diff-dfbead0c52d39f043089365b77b51b9d7e7dca3570c1bcbbbca5b2436efdc21eR249 I feel like `_close_page` will still raise an error if the response is missing

girarda commented:
@maxi297 I updated the PR with the missing piece to skip the page validation if an error was thrown and instead return an AirbyteLogMessage.

I also updated [the error handling which was returning a log message containing the error instead of an AirbyteRecord whose data is a StreamRead containing the log messages](https://github.com/airbytehq/airbyte/pull/24173/commits/b4c45590ad05169767fa1e915ab00079c34b72d4#diff-f655e63774ac8a9d9b3b6f37f553a5e726643dc5da46d39ab11e879775b8a3edR341-R344).

maxi297 commented on airbyte-cdk/python/connector_builder/message_grouper.py:
```diff
@@ -88,7 +88,7 @@ def _get_message_groups(
         while records_count < limit and (message := next(messages, None)):
             if self._need_to_close_page(at_least_one_page_in_group, message):
-                self._close_page(current_page_request, current_page_response, current_slice_pages, current_page_records)
+                self._close_page(current_page_request, current_page_response, current_slice_pages, current_page_records, True)
```
I think this line challenges [this assumption](https://github.com/airbytehq/airbyte/pull/24173/commits/b4c45590ad05169767fa1e915ab00079c34b72d4#diff-01d26032991ab687a8a2142afa745bbb0b92e4cb190835a19c8f3b151a10c699R129). I honestly don't mind too much but it means that we expect he CDK to possibly continue the read on non breaking exceptions  without a response and we should make it clear in the behaviour of this method. I think it translate in removing the param `validate_page_complete` because it is now useless

clnoll commented on airbyte-cdk/python/connector_builder/message_grouper.py:
```diff
@@ -101,33 +103,50 @@ def _get_message_groups(
+            yield from source.read(logger=self.logger, config=config, catalog=configured_catalog, state={})
+        except Exception as e:
+            error_message = f"{e.args[0] if len(e.args) > 0 else str(e)} - {ErrorFormatter.get_stacktrace_as_string(e)}"
+            yield AirbyteMessage(type=MessageType.LOG, log=AirbyteLogMessage(level=Level.ERROR, message=error_message))
```
Out of curiosity, why are we returning a `MessageType.LOG` here instead of `MessageType.TRACE`?

girarda replied:
We're returning it as a log message so it can be [appended to the list of log messages](https://github.com/airbytehq/airbyte/pull/24173/files#diff-01d26032991ab687a8a2142afa745bbb0b92e4cb190835a19c8f3b151a10c699R52).

You bring up a good point though. It might make sense to return a trace message and also add these to the StreamRead's logs

clnoll replied:
Yeah the advantage of that is that it might be nice for the Connector Builder server to parse it as a trace message rather than having to look for errors in a log message type, when deciding what kind of response to send to the client.

clnoll reviewed (approved):
LGTM! Just one question for you @girarda.

brianjlai commented on airbyte-cdk/python/connector_builder/message_grouper.py:
```diff
@@ -101,33 +109,53 @@ def _get_message_groups(
         )
         current_page_records.clear()
 
+    def _read_stream(self, source, config, configured_catalog) -> Iterator[AirbyteMessage]:
```
can we add type hints to this and the `_close_page()` methods

brianjlai reviewed (approved):
non blocking suggestion, but rest looks good!
